core/src/main/scala/scalafix/rewrite/Xor2Either.scala

```diff
@@ -1,28 +1,55 @@
 package scalafix.rewrite
 
+import scala.collection.immutable.{ListMap, Seq}
 import scala.meta.parsers.Parse
 import scala.{meta => m}
-import scalafix.util.Patch
+import scalafix.util._
+import PartialFunction._
+import scala.meta.Type.Select
+import scala.meta._
+
+class Desugared[T <: Tree: Parse](implicit semantic: SemanticApi) {
+  def unapply(original: T): Option[T] = semantic.desugared(original)
+}
 
 case object Xor2Either extends Rewrite {
   override def rewrite(ast: m.Tree, ctx: RewriteCtx): Seq[Patch] = {
-    import scala.meta._
-    val semantic = getSemanticApi(ctx)
-    class Desugared[T <: Tree: Parse] {
-      def unapply(original: T): Option[T] = semantic.desugared(original)
-    }
+    implicit val semanticApi: SemanticApi = getSemanticApi(ctx)
     object DType extends Desugared[Type]
     object DTerm extends Desugared[Term]
-    // NOTE. This approach is super inefficient, since we run semantic.desugar on
-    // every case for every node in the tree. Ideally, we first match on the
-    // syntax structure we want and then run semantic.desugar.
-    ast.collect {
-      case t @ DType(t"cats.data.Xor") =>
-        Patch(t.tokens.head, t.tokens.last, s"Either")
-      case t @ DTerm(q"cats.data.Xor.Right.apply[..$_]") =>
-        Patch(t.tokens.head, t.tokens.last, s"Right")
-      case t @ DTerm(q"cats.data.Xor.Left.apply[..$_]") =>
-        Patch(t.tokens.head, t.tokens.last, s"Left")
-    }
+
+    val typeChanger = new ChangeType(ast, ctx)
+    val methodChanger = new ChangeMethod(ast, ctx)
+    val importAdder = new AddImport(ast, ctx)
+
+    //Create a sequence of type replacements
+    val replacementTypes = List(
+      ReplaceType(t"cats.data.XorT", t"cats.data.EitherT", "EitherT"),
+      ReplaceType(t"cats.data.Xor", t"scala.util.Either", "Either"),
+      ReplaceType(t"cats.data.Xor.Left", t"scala.util.Left", "Left"),
+      ReplaceType(t"cats.data.Xor.Right", t"scala.util.Either.Right", "Right")
+    )
+
+    //Add in some method replacements
+    val replacementTerms = List(
+      ReplaceTerm(t"cats.data.Xor.Right.apply", "Right"),
+      ReplaceTerm(t"cats.data.Xor.Left.apply", "Left")
+    )
+
+    //Then add needed imports.
+    //todo - derive this from patches created, types
+    //and terms replaced
+    //Only add if they are not already imported
+    val additionalImports = List(
+      "cats.data.EitherT",
+      "cats.implicits._",
+      "scala.util.Either"
+    )
+
+    val typeReplacements = typeChanger.gatherPatches(replacementTypes)
+    val termReplacements = methodChanger.gatherPatches(replacementTerms)
+    val addedImports = importAdder.gatherPatches(additionalImports)
+
+    addedImports ++ typeReplacements ++ termReplacements
```

Comment:
This class is now a rough approximation of what I think a nice api for creating type refactorings might look like.  List some types to replace, list some methods to replace, and either derive new imports needed or list those as well.